sw/device/lib/dif/dif_kmac.h

```diff
@@ -0,0 +1,788 @@
+    const char *data, size_t len, dif_kmac_customization_string_t *out);
+
+/**
+ * Encode a function name (N).
+ *
+ * The length of the string must not exceed `kDifKmacMaxFunctionNameLen`.
+ *
+ * Note that this function will encode `len` bytes from `data` regardless of
+ * whether `data` is null-terminated or not.
+ *
+ * @param data String to encode.
+ * @param len Length of string to encode.
+ * @param[out] out Encoded function name.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_string_encode_result_t dif_kmac_function_name_init(
+    const char *data, size_t len, dif_kmac_function_name_t *out);
+
+/**
+ * Start a SHA-3 operation.
+ *
+ * @param kmac A KMAC handle.
+ * @param mode The SHA-3 mode of operation.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_configure_result_t dif_kmac_mode_sha3_start(dif_kmac_t *kmac,
+                                                     dif_kmac_mode_sha3_t mode);
+
+/**
+ * Start a SHAKE operation.
+ *
+ * @param kmac A KMAC handle.
+ * @param mode The mode of operation.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_configure_result_t dif_kmac_mode_shake_start(
+    dif_kmac_t *kmac, dif_kmac_mode_shake_t mode);
+
+/**
+ * Start a cSHAKE operation.
+ *
+ * @param kmac A KMAC handle.
+ * @param mode The mode of operation.
+ * @param S Customization string (optional).
+ * @param N Function name (optional).
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_configure_result_t dif_kmac_mode_cshake_start(
+    dif_kmac_t *kmac, dif_kmac_mode_cshake_t mode,
+    const dif_kmac_customization_string_t *S,
+    const dif_kmac_function_name_t *N);
+
+/**
+ * The result of starting a KMAC operation.
+ */
+typedef enum dif_kmac_mode_kmac_start_result {
+  kDifKmacModeKmacStartOk = kDifKmacConfigureOk,
+  kDifKmacModeKmacStartBadArg = kDifKmacConfigureBadArg,
+  kDifKmacModeKmacStartError = kDifKmacConfigureError,
+  kDifKmacModeKmacStartLocked = kDifKmacConfigureLocked,
+
+  /**
+   * The key did not meet the security requirements for the selected mode.
+   * Please try either selecting a different mode or using a longer key.
+   */
+  kDifKmacModeKmacStartShortKey,
+} dif_kmac_mode_kmac_start_result_t;
+
+/**
+ * Start a KMAC operation.
+ *
+ * To use KMAC in XOF mode set the output length (`L`) to 0.
+ *
+ * The `key` provided must have at least as many bits as the security strength
+ * of the `mode`.
+ *
+ * @param kmac A KMAC handle.
+ * @param mode The mode of operation.
+ * @param L Output length (number of bytes that will be 'squeezed').
+ * @param K Pointer to secret key.
+ * @param S Customization string (optional).
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_kmac_mode_kmac_start_result_t dif_kmac_mode_kmac_start(
+    dif_kmac_t *kmac, dif_kmac_mode_kmac_t mode, size_t L, dif_kmac_key_t *K,
+    const dif_kmac_customization_string_t *S);
```

Comment:
Capital letter arguments? Is this an LLVMism that slipped through?

Reply:
I thought I'd try bending the rules a little bit so that the parameter names would match the spec exactly and see what you guys thought... I'll change them to lower case :)